### ISSUE TITLE
Reproduce the results from the paper.

### ISSUE THREAD

**Issue by going-doer** [atutej/RARe#1]
Following the instructions in the README, I attempted to reproduce the results from the paper, specifically using the E5-Mistral-Instruct as the base model. Unfortunately, I couldn't achieve results similar to those reported in the paper.

1. The evaluation instructions don't seem to work for me ([link](https://github.com/atutej/RARe?tab=readme-ov-file#32-evaluation)). The provided script (run_eval.sh) includes the model name, but I’m unable to load the model.

2. As a workaround, I trained the model and then attempted the evaluation. I modified run_eval.py to load the trained model. However, the results still differ significantly from the paper's reported performance. For example, on the Arguana dataset with 5-shot evaluation, I obtained a main score of 0.22095, whereas the paper reports 60.87.

I was wondering if you have any plans to release the pre-trained models to support reproducibility. 
Additionally, could you provide any guidance or suggestions on reproducing the results from the paper?

If you need more detailed information, feel free to let me know.

Thank you for sharing the code and for all your hard work!

**Comment by atutej** [comment]
Hi, thank you for your interest in our work!

The pre-trained models are indeed already available on HF, for E5-Mistral-Instruct that would be: https://huggingface.co/atutej/RARe-E5-Mistral-7B-Instruct-Ret. You can load it with the name defined in this line [here](https://github.com/atutej/RARe/blob/1cc15313d6c4ac1635d4398eba5ab3de50a33297/mteb/mteb/models/e5_instruct.py#L98)). 
i.e. command: `CUDA_VISIBLE_DEVICES=0 python3 run_eval.py --model_name_or_path RARe-E5-Mistral-7B-Instruct-Ret --task "ArguAna" --n_ic_examples 5 --batch_size 64`

I ran evaluation with the above command, and get the following numbers for ArguAna:

> "main_score": 0.60854,
        "map_at_1": 0.35917,
        "map_at_10": 0.51967,
        "map_at_100": 0.52635,
        "map_at_1000": 0.52637,
        "map_at_20": 0.52565,
        "map_at_3": 0.47202,
        "map_at_5": 0.49912,
        "mrr_at_1": 0.37197724039829305,
        "mrr_at_10": 0.5243014631172539,
        "mrr_at_100": 0.5310431953170244,
        "mrr_at_1000": 0.531063502787952,
        "mrr_at_20": 0.5303464839618517,
        "mrr_at_3": 0.4761735419630163,
        "mrr_at_5": 0.5035917496443826,

Could you please share the error log/command that you used to run the evaluation mentioned in point 1? (edit: I believe the examples in run_eval.sh had an extra "-hf" suffix in the model name which does not exist in the actual path, which should be fixed now!)

For point 2, could you please share the modifications that you made? Ideally, you would only need to change https://github.com/atutej/RARe/blob/main/mteb/mteb/models/e5_instruct.py to include the path to your newly trained model.

**Comment by going-doer** [comment]
Thank you for your quick response!


### 1. [RESOLVED] The evaluation instructions don't seem to work for me. 

I got the same result you mentioned when using the pre-trained model. Thank you for your help!

### 2. I trained the model and then attempted the evaluation.

I was confused by the instructions; it was my mistake.
Since the training seems to be based on llm2vec, I initially referred to llm2_vec.py instead of e5_instruct.py. 
Following your comment, I switched to referencing e5_instruct.py.
However, the performance of my trained model (5-shot) on ArguAna is still quite low (0.57793).
Could this be the cause of the difference in performance?

---

I also have three additional questions regarding the training process:

### 3. Could you please check the training code?

When running the script, I encountered an error in E5Data.py at line 128 ([link](https://github.com/atutej/RARe/blob/06fdb3f9b2601e9883783b63680a22ae7b1b8c05/llm2vec/llm2vec/dataset/E5Data.py#L129)).
It seems that construct_ic_query returns only one value ([link](https://github.com/atutej/RARe/blob/06fdb3f9b2601e9883783b63680a22ae7b1b8c05/llm2vec/llm2vec/dataset/dataset.py#L92)).
To address this, I modified the code as follows:

```python
                # new_query, new_query_flipped = self.construct_ic_query(sample, self.n_ic_examples, searcher=searcher, ic_data=separated_samples[dataset], use_negatives=self.use_negatives, instruction=instruction if self.use_instructions else None)
                new_query = self.construct_ic_query(sample, self.n_ic_examples, searcher=searcher, ic_data=separated_samples[dataset], use_negatives=self.use_negatives, instruction=instruction if self.use_instructions else None)
                new_query_flipped = new_query 
```

Is this the correct modification? Could you please check the training code?


### 4. Is using `llm2vec` the correct way to train `intfloat/e5-mistral-7b-instruct`?

While evaluating the model "RARe-E5-Mistral-7B-Instruct-Ret", the code seems to use gritlm instead of llm2vec ([link](https://github.com/atutej/RARe/blob/06fdb3f9b2601e9883783b63680a22ae7b1b8c05/mteb/mteb/models/e5_instruct.py#L23)).

However, the training code utilizes llm2vec even when starting from "intfloat/e5-mistral-7b-instruct" ([link](https://github.com/atutej/RARe/blob/main/llm2vec/train_configs/supervised/E5-Instruct.json)).

Is using llm2vec the correct way to train intfloat/e5-mistral-7b-instruct?
If so, why does the evaluation code use gritlm?


### 5. The training loss remains stuck at zero when using `McGill-NLP/LLM2Vec-Meta-Llama-3-8B-Instruct-mntp-supervised` as the base model.

When using `McGill-NLP/LLM2Vec-Meta-Llama-3-8B-Instruct-mntp-supervised` as the base model, the training loss remains at zero.
I followed the configuration in llm2vec_models.py ([link](https://github.com/atutej/RARe/blob/06fdb3f9b2601e9883783b63680a22ae7b1b8c05/mteb/mteb/models/llm2vec_models.py#L234)) and trained the model using the "MetaLlama3-Supervised.json" config ([link](https://github.com/atutej/RARe/blob/main/llm2vec/train_configs/supervised/MetaLlama3-Supervised.json)).
Similar to before, I only adjusted "per_device_train_batch_size", "per_device_eval_batch_size", and "gradient_accumulation_steps" to (8, 8, 16) to fit my GPU.
Do you have any insights on why the training loss is stuck at zero?


If you need any additional information, please feel free to let me know!  
Thank you again for your time and support.

**Comment by atutej** [comment]
Good to know 1. is resolved!

> 2.  I trained the model and then attempted the evaluation.

If you are using the modified `per_device_batch_size` from comment 5, that is likely the reason. At least for contrastive learning, the higher the `per_device_batch_size`, the better (eg: https://openreview.net/forum?id=T1dhAPdS--). The pre-trained model was run with `per_device_batch_size 32`, `grad_accumulation_steps 2` on 8 H100 GPUs. Decreasing per_device_batch_size and increasing gradient_accumulation_steps unfortunately does not have the same effect for contrastive loss because the loss for each example depends on every other example in the mini-batch. I ran E5-Mistral-Instruct with RARe (5 ic examples) on 4 H100 GPUs with `per_device_batch_size 32`, `grad_accumulation_steps 2`, and obtain the following numbers on ArguAna:

```
"main_score": 0.60693,
        "map_at_1": 0.33784,
        "map_at_10": 0.51209,
        "map_at_100": 0.51781,
        "map_at_1000": 0.51783,
        "map_at_20": 0.51723,
        "map_at_3": 0.46207,
        "map_at_5": 0.49247,
        "mrr_at_1": 0.35135135135135137,
        "mrr_at_10": 0.5171896520580743,
        "mrr_at_100": 0.5227731201584701,
        "mrr_at_1000": 0.5227916606094393,
        "mrr_at_20": 0.5221884927196984,
        "mrr_at_3": 0.4669274537695595,
        "mrr_at_5": 0.49719061166429707,
```

If you are limited by batch size, a workaround is to use GradCache which is specifically designed for contrastive loss: https://github.com/luyug/GradCache, however I wasn't able to successfully set that up for my initial experiments with the LLM2Vec training scripts.

> 3. When running the script, I encountered an error in E5Data.py at line 128 

Thank you for catching this! Your workaround should work perfectly fine because `query_flipped` is not used anyway. I will fix these soon and upload a new commit.

> 4. Is using llm2vec the correct way to train intfloat/e5-mistral-7b-instruct?

Unfortunately the original E5-Mistral-Instruct training code is not publicly released, so we do not know what the best way of training the model is. We adapt LLM2Vec training script because it provided a seamless way to train on a dataset mixture that was closest to what is mentioned in the original E5-Mistral Paper. Indeed, we had to modify some aspects, such as using the "Instruct: {instruct} Query: {query}" format mentioned in the [Readme for E5-Mistral-Instruct](https://huggingface.co/intfloat/e5-mistral-7b-instruct), while LLM2Vec uses a slightly different format. We also had to change to causal attention and last token pooling.

The evaluation code is directly adapted from the original MTEB code as it uses the GritLM wrapper for E5-Mistral-Instruct, even though the original training script is not known. To the best of my understanding, the GritLM wrapper in MTEB is a general purpose wrapper that handles the additional instruction format, which is also used by E5-Mistral-Instruct, and is not strictly for models trained with GritLM training scripts. I recommend using the `e5_instruct.py` for all E5-Mistral-Instruct variants for this reason. 

5. I have not encountered this situation before. I will try training `McGill-NLP/LLM2Vec-Meta-Llama-3-8B-Instruct-mntp-supervised` to see if I can reproduce your issue and follow up soon. 

Hope this has clarified some of your questions!

**Comment by going-doer** [comment]
Thank you for your quick and kind response, and for providing helpful suggestions!  
Questions 1, 2, 3, and 4 have been resolved.  

For question 5, I will try training `McGill-NLP/LLM2Vec-Meta-Llama-3-8B-Instruct-mntp-supervised` again using [GradCache](https://github.com/luyug/GradCache). 
If I encounter the same issue, I will reopen the issue with more detailed information.  
If you experience the same zero-loss issue, please feel free to reopen this issue.  

I truly appreciate your support. Thank you! :)

**Comment by atutej** [comment]
I tried running LLM2Vec with your configuration, but I was not able to replicate your zero-loss issue. I have attached my training log so far and also my current requirements and environment information, in case there is a mismatch in some package versions.

[llm2vec_log.txt](https://github.com/user-attachments/files/18381140/llm2vec_log.txt)
[current_env.txt](https://github.com/user-attachments/files/18381176/current_env.txt)

**Comment by going-doer** [comment]
I recreated the environment (rare_llama3_env.txt) similar to the attached files (current_env.txt).
After that, the loss started decreasing as expected, similar to what you shared. 
However, at epoch 0.62, the loss suddenly spiked and then got stuck at zero. (rare_llama3_log.txt)

Before changing the environment, this issue occurred earlier at epoch 0.21, as shown in the log file I attached (rare_llama3_log_old.txt).

But, In the same environment, training with intfloat/e5-mistral-7b-instruct works properly. 
I'm not sure why this issue is happening.

For the GPU setup, I ran the experiment on 4× A6000 (48GB) GPUs using the attached configuration(MetaLlama3-Supervised.json), where I only modified per_device_train_batch_size, per_device_eval_batch_size, and gradient_accumulation_steps.

[rare_llama3_env.txt](https://github.com/user-attachments/files/18404553/rare_llama3_env.txt)
[rare_llama3_log.txt](https://github.com/user-attachments/files/18404555/rare_llama3_log.txt)
[rare_llama3_log_old.txt](https://github.com/user-attachments/files/18404554/rare_llama3_log_old.txt)
[MetaLlama3-Supervised.json](https://github.com/user-attachments/files/18404552/MetaLlama3-Supervised.json)

If you have any ideas on how to resolve this issue, please let me know! 
I will also continue investigating this problem.

Thank you for your help!

**Comment by atutej** [comment]
Thanks for sharing your log. 

My run on 4H100s (almost) completed (it was terminated at the 192/194 step due to time limit) but the loss continued to stay stable, attached below. Also attached my configuration. 

[MetaLlama3-Supervised.json](https://github.com/user-attachments/files/18406062/MetaLlama3-Supervised.json)
[full_rare_llm2vec.txt](https://github.com/user-attachments/files/18406063/full_rare_llm2vec.txt)

From what I see in your log, your grad norm is going to "nan" at 0.62, causing loss to spike and then go to 0. I am not sure why this could be happening, here are some steps maybe you can try to debug: 

- It might be useful to debug the loss (and inputs to the loss) when this happens.
- Since the batch size is quite small, there might be a single negative example that is causing the loss to go to nan.
- You could also try changing dtype to float32 and/or disable flash attention.
